llpc/lower/llpcSpirvLowerMath.cpp

```diff
@@ -295,16 +283,154 @@ bool SpirvLowerMathPrecision::runImpl(Module &module) {
         valueWritten = callInst->getOperand(0);
       }
 
-      if (valueWritten && builtIn == lgc::BuiltInPosition && enableImplicitInvariantExports) {
+      if (valueWritten && builtIn == lgc::BuiltInPosition) {
         disableFastMath(valueWritten);
         changed = true;
       }
     }
   }
+  return changed;
+}
+
+static bool clearContractFlag(Instruction *inst) {
+  if (!isa<FPMathOperator>(inst))
+    return false;
+  LLVM_DEBUG(dbgs() << "clearing contract flags: " << *inst << "\n");
+  FastMathFlags fastMathFlags = inst->getFastMathFlags();
+  fastMathFlags.setAllowReassoc(false);
+  fastMathFlags.setAllowContract(false);
+  inst->copyFastMathFlags(fastMathFlags);
+  return true;
+}
+
+bool SpirvLowerMathPrecision::propagateNoContract(Module &module, bool forward, bool backward) {
+  bool changed = false;
+
+  SmallVector<Instruction *> roots;
+  DenseSet<Instruction *> visited;
+
+  // Find all NoContract instructions to build root set
+  LLVM_DEBUG(dbgs() << "locate no contract roots\n");
+  for (auto &func : module) {
+    for (auto &block : func) {
+      for (auto &inst : block) {
+        if (isNoContract(&inst)) {
+          LLVM_DEBUG(dbgs() << "root: " << inst << "\n");
+          roots.push_back(&inst);
+          visited.insert(&inst);
+        }
+      }
+    }
+  }
+
+  SmallVector<Instruction *> worklist;
+
+  // Backward propagate via operands
+  if (backward) {
+    LLVM_DEBUG(dbgs() << "backward propagate no contract\n");
+    worklist = roots;
+    while (!worklist.empty()) {
+      auto inst = worklist.pop_back_val();
+      LLVM_DEBUG(dbgs() << "visit: " << *inst << "\n");
+      for (Value *operand : inst->operands()) {
+        if (auto opInst = dyn_cast<Instruction>(operand)) {
+          if (visited.count(opInst))
+            continue;
+          if (clearContractFlag(opInst))
+            changed = true;
+          worklist.push_back(opInst);
+          visited.insert(opInst);
+        }
+      }
+    }
+  }
+
+  // Forward propagate via users
+  if (forward) {
+    LLVM_DEBUG(dbgs() << "forward propagate no contract\n");
+    worklist = roots;
+    while (!worklist.empty()) {
+      auto inst = worklist.pop_back_val();
+      LLVM_DEBUG(dbgs() << "visit: " << *inst << "\n");
+      for (User *user : inst->users()) {
+        // Only propagate through instructions
+        if (auto userInst = dyn_cast<Instruction>(user)) {
+          if (visited.count(userInst))
+            continue;
+          // Only update FAdd instructions
+          if (userInst->getOpcode() == Instruction::FAdd) {
+            if (clearContractFlag(userInst))
+              changed = true;
+          }
+          worklist.push_back(userInst);
```

Comment:
Do we really need to propagate further than the first `fadd`?

Reply:
Yes, I believe we do if our goal is to replicate the previous code.

The previous code ran in an instruction visitor and as such it could strip _contract_ from one FAdd (a), then find that FAdd (a) as a input to a later FAdd (b).
With the caveat that the instruction visitor progresses in program order, so it is possible that (b) could be reached before (a), yielding a different result depending on program structure.
This implementation works independent of program order.

That said, I am not against changing this to first FAdd only -- if we are happy in the change in behaviour.